apps/state/lib/state/trip.ex

```diff
@@ -182,35 +162,45 @@ defmodule State.Trip do
     [new_trip | new_alternates]
   end
 
-  defp do_apply_filters(filters, trips \\ [])
+  defp do_apply_filters(%{routes: []}), do: []
+  defp do_apply_filters(%{route_patterns: []}), do: []
+  defp do_apply_filters(%{ids: []}), do: []
+  defp do_apply_filters(%{direction_id: _id} = filters) when map_size(filters) == 1, do: []
+  defp do_apply_filters(%{date: _date} = filters) when map_size(filters) == 1, do: []
 
-  defp do_apply_filters(%{routes: route_ids} = filters, trips) do
-    direction_id = filters[:direction_id]
-
-    filtered_trips =
-      for route_id <- route_ids do
-        by_route_and_direction_id(route_id, direction_id)
-      end
+  defp do_apply_filters(filters) do
+    matchers =
+      []
+      |> build_filters(:route_id, filters[:routes])
+      |> build_filters(:direction_id, filters[:direction_id])
+      |> build_filters(:route_pattern_id, filters[:route_patterns])
+      |> build_filters(:id, filters[:ids])
 
-    new_trips =
-      filtered_trips
-      |> List.flatten()
-      |> Enum.concat(trips)
+    trips = State.Trip.select(matchers) ++ State.Trip.Added.select(matchers)
 
-    filters
-    |> Map.drop([:routes, :direction_id])
-    |> do_apply_filters(new_trips)
+    case filters[:date] do
+      nil -> trips
+      date -> Enum.filter(trips, &ServiceByDate.valid?(&1.service_id, date))
+    end
   end
 
-  defp do_apply_filters(%{date: date = %Date{}} = filters, trips) do
-    new_trips = Enum.filter(trips, &ServiceByDate.valid?(&1.service_id, date))
+  defp build_filters(matchers, _key, nil), do: matchers
 
-    filters
-    |> Map.delete(:date)
-    |> do_apply_filters(new_trips)
+  defp build_filters(matchers, key, values) when is_list(values) do
+    if matchers == [] do
```

Comment:
If you start with `matchers = [%{}]` when you kick off these calls, you don't need the extra `if` case.

Reply:
You could also do
```elixir
values = List.wrap(values)
```
and then you don't need to handle the separate case where `values` is a single item.